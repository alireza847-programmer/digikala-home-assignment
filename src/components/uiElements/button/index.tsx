import React from 'react';
import Button from './style';
import {DButtonProps} from 'types/components/uiElements/button';
import DText from '../text';
import {theme} from 'themes/emotion';
import DRow from '../row';

const DButton = (props: DButtonProps) => {
  const {
    mode = 'primary',
    subTitle,
    title,
    styled = 'FULL_WIDTH',
    icon,
    rightIcon,
    onPress,
    testID,
    disabled = false,
    width,
  } = props;

  const renderContentColor = () => {
    if (disabled) {
      return theme.colors.textColor.button.disabled;
    }
    if (styled === 'TEXT') {
      return theme.colors.button[mode];
    }
    return theme.colors.textColor.button[mode];
  };

  return (
    <Button
      width={width}
      disabled={disabled}
      testID={testID}
      activeOpacity={0.8}
      onPress={onPress}
      styled={styled}
      mode={mode}>
      {icon && (
        <DRow {...(title && {marginRighRatio: 2})} fullWidth={false}>
          {icon(renderContentColor())}
        </DRow>
      )}
      <DText
        typography="semiBold16"
        color={renderContentColor()}
        marginTopRatio={0}>
        {title}
      </DText>
      {subTitle && (
        <DText
          typography="regular16"
          color={renderContentColor()}
          marginLeftRatio={1.2}
          marginTopRatio={0}>
          {subTitle}
        </DText>
      )}
      {rightIcon && (
        <DRow {...(title && {marginLeftRatio: 2})} fullWidth={false}>
          {rightIcon(renderContentColor())}
        </DRow>
      )}
    </Button>
  );
};
export default DButton;
