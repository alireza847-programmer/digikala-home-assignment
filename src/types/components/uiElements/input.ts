import {TextInputProps} from 'react-native';

export interface InputProps {
  marginTopRatio?: number;
}

export interface DInputProps extends InputProps, TextInputProps {}
