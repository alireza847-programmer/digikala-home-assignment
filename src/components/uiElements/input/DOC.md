# DInput Component

## Overview

The `DInput` component is a customizable input field designed for use in React Native applications. It provides a clean and consistent interface for capturing user input.

## Style

The styling for the `DInput` component is defined in the `Input` module. It uses Emotion's styling system for enhanced readability and maintainability.

```jsx
import styled from '@emotion/native';
import {InputProps} from 'types/components/uiElements/input';

const DInput =
  styled.TextInput <
  InputProps >
  (({theme, marginTopRatio = 1}) => ({
    borderWidth: 1,
    width: '100%',
    padding: theme.HorizontalRem * 4,
    borderRadius: theme.radius * 2,
    borderColor: theme.colors.stroke.input,
    fontFamily: theme.fonts.medium15.fontFamily,
    fontSize: theme.fonts.medium15.fontSize,
    marginTop: theme.verticalRem * marginTopRatio,
  }));

export default DInput;
```

## Usage

```jsx
import React, {useState} from 'react';
import {View, Text} from 'react-native';
import DInput from './path-to-your-component/DInput';

const ExampleScreen = () => {
  const [inputValue, setInputValue] = useState('');

  const handleInputChange = (text: string) => {
    setInputValue(text);
  };

  return (
    <View>
      <Text>Enter your text:</Text>
      <DInput
        value={inputValue}
        onChangeText={handleInputChange}
        placeholder="Type here..."
        marginTopRatio={2}
      />
    </View>
  );
};

export default ExampleScreen;
```

## Props

### DInput Props

- `marginTopRatio` (number): The ratio to set the top margin for the input.

### Input Props

The `DInput` component accepts standard `TextInputProps` from React Native.
