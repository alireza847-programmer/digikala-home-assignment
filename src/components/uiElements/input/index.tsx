import React from 'react';
import Input from './style';
import {DInputProps} from 'types/components/uiElements/input';
import {theme} from 'themes/emotion';

const DInput = (props: DInputProps) => {
  const {...rest} = props;
  return (
    <Input
      placeholderTextColor={theme.colors.textColor.placeHolder}
      {...rest}
    />
  );
};

export default DInput;
