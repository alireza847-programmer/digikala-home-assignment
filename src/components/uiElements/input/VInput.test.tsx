import React from 'react';
import {render, fireEvent} from '@testing-library/react-native';
import DInput from './index';
import {it, describe, expect} from '@jest/globals';
import {DInputProps} from 'types/components/uiElements/input';
import {ThemeProvider} from '@emotion/react';
import {theme} from 'themes/emotion';

const Component = (props: DInputProps) => (
  <ThemeProvider theme={theme}>
    <DInput {...props} />
  </ThemeProvider>
);

describe('DInput component', () => {
  it('renders correctly with default props', () => {
    const {toJSON} = render(<Component />);
    expect(toJSON()).toMatchSnapshot();
  });

  it('renders correctly with custom marginTopRatio', () => {
    const {toJSON} = render(<Component marginTopRatio={2} />);
    expect(toJSON()).toMatchSnapshot();
  });
});
