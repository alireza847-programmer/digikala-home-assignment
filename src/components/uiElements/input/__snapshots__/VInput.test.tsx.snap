// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DInput component renders correctly with custom marginTopRatio 1`] = `
<TextInput
  marginTopRatio={2}
  placeholderTextColor="#A3ACB8"
  style={
    [
      {
        "borderColor": "#0071F3",
        "borderRadius": 15,
        "borderWidth": 1,
        "fontFamily": "Inter-Medium",
        "fontSize": 15,
        "marginTop": 13.34,
        "padding": 30,
        "width": "100%",
      },
      undefined,
    ]
  }
/>
`;

exports[`DInput component renders correctly with default props 1`] = `
<TextInput
  placeholderTextColor="#A3ACB8"
  style={
    [
      {
        "borderColor": "#0071F3",
        "borderRadius": 15,
        "borderWidth": 1,
        "fontFamily": "Inter-Medium",
        "fontSize": 15,
        "marginTop": 6.67,
        "padding": 30,
        "width": "100%",
      },
      undefined,
    ]
  }
/>
`;
